### ISSUE TITLE
Implement File class

### ISSUE THREAD

**Issue by jwosty** [jwosty/Fable.System.IO#3]
It would be pretty cool to make calling File.Open work in some circumstances. Here would be the ideal approach:

The File class would be implemented in such a way that it would accept URIs as paths (and would never allow writing; only reading), so you could do something like this:

```f#
use contents = File.OpenText("https://example.com/something.json")
let result = contents.ReadToEnd ()
printfn "something.json:\n%s" result
```

It should also be able to handle relative paths, which would be treated as if they are relative to some kind of easy-to-reason about point -- perhaps the top-level uri of the host website. In other words, say we have the following website structure:

*  https://example.com/
    * app.js
    * something.json

Calling ``File.OpenText("something.json")`` from within app.js should obviously return the contents from ``https://example.com/something.json``. The only question would be: what is this "current directory" actually relative to? Is it the website root? Or is it relative to the current script? In other words, in the following scenario, which ``something.json`` should we point to?

* https://example.com/
    * something.json
    * app/
        * something.js
        * app.js

One final note -- this really is exactly equivalent to setting some reasonable environment-wide CurrentDirectory, while extending all these APIs to understand URIs instead of paths. Perhaps the current directory should be settable as well, like it normally is (provided that you give it a URI instead?).

**Comment by jwosty** [comment]
This could potentially allow a big swath of F# libraries to easily gain Fable compatibility, one example being the [ResX type provider](https://fsprojects.github.io/FSharp.Configuration/ResXProvider.html), which inherently reads its data from files.

**Comment by Shmew** [comment]
The way you'd want to handle reading the files (and what I usually do) is just make a GET request for that resource. You can see an example of how this is done using [Fable.SimpleHttp](https://github.com/Zaid-Ajaj/Fable.SimpleHttp) in an example for one of my libraries [here](https://github.com/Shmew/Feliz.Plotly/blob/master/docs/Plotly/Examples/Statistical/Splom/Iris.fs). One thing you'll want to keep in mind is if the user is in a node.js environment in which case you can actually just read the file using the [file system API](https://nodejs.org/api/fs.html#fs_file_system). 

I'd recommend either making the user specify the full path themselves or doing `document.location.href` and stripping any hash routing for the browser. In node.js environments you could just do `__dirname` as the default path.

I'd also suggest adding support for writing files using that same node.js file system API. You could even potentially implement writing files for browser environments by doing a POST and letting a user specify an endpoint, but that kind of stuff is tricky enough that I'd probably not bother.

**Comment by jwosty** [comment]
> One thing you'll want to keep in mind is if the user is in a node.js environment

I thought about the node.js option too -- is it a frequently used deployment model for Fable code? I _assumed_ it wouldn't be, since you'd just use dotnet to run F# on the server side, and that 9 times out of 10 you'd be deploying Fable code to the browser. In which case, if there's not a huge demand to have it working on node, then I might not bother (it would also have the cost of me having to learn Node's APIs and study them :) ). Law of diminishing returns, perhaps. Though perhaps there's a case to be made for Electron apps built with Fable, as I do believe those use Node, correct?

> I'd recommend either making the user specify the full path themselves or doing document.location.href and stripping any hash routing for the browser. In node.js environments you could just do __dirname as the default path.

Seems sensible. But wouldn't it be really nice if ``File.OpenText("foo.json")`` "just worked" in most situations on both dotnet and a Fable browser app? I.e. no conditional compilation needed there!

> I'd also suggest adding support for writing files using that same node.js file system API

I agree; if node.js is supported, then File writing should definitely work in that scenario.

> You could even potentially implement writing files for browser environments by doing a POST and letting a user specify an endpoint, but that kind of stuff is tricky enough that I'd probably not bother.

Right, I probably wouldn't bother with POST requests as this starts to get into territory that I would want the user to really know what they're doing if they want to open up that route. I don't think I'd want to encourage usage of an API like that, as it would open up a big, easy attack vector (just flood the server with malicious file writes using that API to fill up its hard drive space -- essentially accomplishing a DDOS). If we did this, it would be an opt-in feature.

**Comment by jwosty** [comment]
Another question would be how to support the async methods. It would be ideal to be able to use them, but that opens a whole other can of worms as they all use Task<T>. Is there a library that supplies this for Fable?

**Comment by Shmew** [comment]
> I thought about the node.js option too -- is it a frequently used deployment model for Fable code?

Yeah very few people are actually targeting node.js outside the few people using electron. I can confirm that there are people using electron though. Up to you if you want to support that, I just figured I'd throw in my two cents.

> Seems sensible. But wouldn't it be really nice if File.OpenText("foo.json") "just worked" in most situations on both dotnet and a Fable browser app? I.e. no conditional compilation needed there!

Yeah, that would be the behavior if you do what I suggested with the hash routing stripping and taking the href.

> Right, I probably wouldn't bother with POST requests as this starts to get into territory that I would want the user to really know what they're doing if they want to open up that route. I don't think I'd want to encourage usage of an API like that, as it would open up a big, easy attack vector (just flood the server with malicious file writes using that API to fill up its hard drive space -- essentially accomplishing a DDOS). If we did this, it would be an opt-in feature.

I agree it's totally not worth it, but figured it was worth mentioning.

> Another question would be how to support the async methods. It would be ideal to be able to use them, but that opens a whole other can of worms as they all use Task. Is there a library that supplies this for Fable?

Task isn't supported, and there is no intention to support it in the future either. I personally would like to see task support simply because it enables a lot of interoperability, but it does increase the maintenance surface area for the compiler guys.

**Comment by jwosty** [comment]
> Task isn't supported, and there is no intention to support it in the future either. I personally would like to see task support simply because it enables a lot of interoperability, but it does increase the maintenance surface area for the compiler guys.

Right; and the unfortunate consequence of that is that you'd be forced to use System.IO.File stuff synchronously. It would be super ideal if something like this would compile:

```
async {
    let! result = Async.AwaitTask (File.ReadAllTextAsync("somefile"))
    // ... more code
}

Perhaps there could be a Task<T> type that just wraps an F# async, and then Async.AwaitTask just unwraps and starts it? The goal being that you wouldn't have to change your real .NET code at all in order to make it work under Fable. Though perhaps that is no so realistic, especially after examining this situation.

**Comment by Shmew** [comment]
Maybe @alfonsogarciacaro can give some input on how we could handle this

**Comment by alfonsogarciacaro** [comment]
Well, you know my opinion, shimming . NET APIs is full of hurdles like this :) I did try to implement File.ReadLines in node.js and had issues (if I don't cheat and just read the whole file and split it in lines) because the readline API in node works with events. If cross-platform support is really needed, for me it's usually better to have a common (and smaller) API that works both for Fable and . NET. Fable.System.IO is a great opportunity for this, for example, by offering methods using F# async instead of Task.

**Comment by Shmew** [comment]
Yeah that's a great point. I do find it hard to think of a situation where I'd need to be reading files that's shared between .net and fable. I do think something like with FAKE's `Fake.IO.FileSystem` is a really nice API that shows how that could be done.

On the other hand it would be really nice if we had Task (and IObservable) support because it would make things like [AsyncSeq](https://github.com/fsprojects/FSharp.Control.AsyncSeq/) natively supported instead of the mess of stuff I had to do in there that only enabled partial API support.

**Comment by alfonsogarciacaro** [comment]
`Observable` is [already supported](https://github.com/fable-compiler/Fable/blob/nagareyama/tests/Main/ObservableTests.fs) :) (and `Event`!). About Task, there's a discussion about it somewhere, but basically JS is single threaded so shimming anything to work with threads is deceiving. `Async` in Fable is just a convenient way to work with callbacks, and this already creates confusion in some users for example when they try to use `Async.RunSynchronously`. I can only see more confusion arising if we try to shim `Task` where the threading semantics is more prevalent.
